waspc/src/Wasp/Psl/Parser/Model.hs

```diff
@@ -33,177 +38,75 @@ import qualified Wasp.Psl.Ast.Model as Model
+
+modelElement :: Parser Psl.Ast.Element
+modelElement =
+  try (Psl.Ast.ElementField <$> modelField)
+    <|> try (Psl.Ast.ElementBlockAttribute <$> blockAttribute)
+
+modelField :: Parser Psl.Ast.Field
+modelField = do
+  name <- identifier
   type' <- fieldType
   maybeTypeModifier <- fieldTypeModifier
-  attrs <- many (try attribute)
+  attrs <- many (try fieldAttribute)
   return $
-    Model.Field
-      { Model._name = name,
-        Model._type = type',
-        Model._typeModifiers = maybeToList maybeTypeModifier,
-        Model._attrs = attrs
+    Psl.Ast.Field
+      { Psl.Ast._name = name,
+        Psl.Ast._type = type',
+        Psl.Ast._typeModifiers = maybeToList maybeTypeModifier,
+        Psl.Ast._attrs = attrs
       }
   where
-    fieldType :: Parser Model.FieldType
+    fieldType :: Parser Psl.Ast.FieldType
     fieldType =
+      scalarFieldType
+        <|> try
+          ( Psl.Ast.Unsupported
+              <$> ( symbol "Unsupported"
+                      >> parens stringLiteral
+                  )
+          )
+        <|> Psl.Ast.UserType <$> identifier
+
+    scalarFieldType :: Parser Psl.Ast.FieldType
+    scalarFieldType =
       foldl1
         (<|>)
         ( map
-            (\(s, t) -> try (T.symbol lexer s) >> return t)
-            [ ("String", Model.String),
-              ("Boolean", Model.Boolean),
-              ("Int", Model.Int),
-              ("BigInt", Model.BigInt),
-              ("Float", Model.Float),
-              ("Decimal", Model.Decimal),
-              ("DateTime", Model.DateTime),
-              ("Json", Model.Json),
-              ("Bytes", Model.Bytes)
+            -- Supported scalar types from https://github.com/prisma/prisma-engines/blob/main/psl/parser-database/src/types.rs#L1429
+            (\(s, t) -> try (reserved s) >> return t)
+            [ ("String", Psl.Ast.String),
+              ("Boolean", Psl.Ast.Boolean),
+              ("Int", Psl.Ast.Int),
+              ("BigInt", Psl.Ast.BigInt),
+              ("Float", Psl.Ast.Float),
+              ("Decimal", Psl.Ast.Decimal),
+              ("DateTime", Psl.Ast.DateTime),
+              ("Json", Psl.Ast.Json),
+              ("Bytes", Psl.Ast.Bytes)
             ]
         )
-        <|> try
-          ( Model.Unsupported
-              <$> ( T.symbol lexer "Unsupported"
-                      >> T.parens lexer (T.stringLiteral lexer)
-                  )
-          )
-        <|> Model.UserType <$> T.identifier lexer
 
     -- NOTE: As is Prisma currently implemented, there can be only one type modifier at one time: [] or ?.
-    fieldTypeModifier :: Parser (Maybe Model.FieldTypeModifier)
+    fieldTypeModifier :: Parser (Maybe Psl.Ast.FieldTypeModifier)
     fieldTypeModifier =
       optionMaybe
-        ( (try (T.brackets lexer (T.whiteSpace lexer)) >> return Model.List)
-            <|> (try (T.symbol lexer "?") >> return Model.Optional)
+        ( (try (symbol "[]?") >> return Psl.Ast.UnsupportedOptionalList)
```

Comment:
Ok, so I would say I didn't confuse what the behaviour is, I just haven't considered this high level behaviour at all.

This is a valid point that you are raising. We are now discussing how we are doing error recovery in our parser.

Usual tactic is to report an error, but then move up to the point where you know you think you can likely recover (error boundary), usually that is some type of declaration, e.g. block in this case, and then continue parsing. So you would loose that one block, but you go to on, and end up with some stuff parsed + some errors, and then consumer of parser can choose what to do with that.

Another way could be to do what you did here, which is encode errors directly into AST as part of it. Well, it can be complementary to the approach above even I believe. But, you are encoding only this specific mistake -> what about all the other errors that could happen? What if there is unexpected `;` between two blocks? Will you be able to handle that? From what seems to me, we don't really have proper error recovery implemented here, you only have a way to capture error fro this specific case, and for anything else you will again get this unwanted behaviour you described where you parse only part of whole prisma.schema and at the end really never know you parsed only part of it.

So, as I said, one solution is to implement proper error recovery with error boundaries. That is actually not super hard, we could do it, but I wouldn't bother with it now as I don't think we really need it. We just need to know if whole parsing was successful or not -> if it wasn't, we can just fail. How can you do that? Well, you can just check, after calling the parser, if anything was left (make sure to add last part to parser which consumes whitespace though). If anything was left, that means not all could be parsed, and you know we failed for our use case, and can just fail the whole thing in the appropriate manner. You don't have a very good way to say where it failed though, for that we need more sophisticated mechanisms, but we can just run prisma format on it if needed and get its output, or they can check it in their IDE or whatever. And we can always improve this later if needed.

Reply:
> From what seems to me, we don't really have proper error recovery implemented here, you only have a way to capture error fro this specific case

To point out why: because Prisma has it in their grammar https://github.com/prisma/prisma-engines/blob/ae0473d812da9ce31f8c62740c937afb9623f2dc/psl/schema-ast/src/parser/datamodel.pest#L55C1-L55C11

> We just need to know if whole parsing was successful or not -> if it wasn't, we can just fail. How can you do that? Well, you can just check, after calling the parser, if anything was left (make sure to add last part to parser which consumes whitespace though). If anything was left, that means not all could be parsed, and you know we failed for our use case, and can just fail the whole thing in the appropriate manner.

This is true! I'll try it out and see that we get reliable failing for the case of "I've gotten unexpected stuff, I didn't parse everything"

Reply:
Now the Prisma schema parser fails well :) as I mentioned in a related comment